// @nuxthq/ui/dist/runtime/components/forms/FormGroup.mjs
import { h, cloneVNode, computed, defineComponent, provide, inject } from "vue";
import { defu } from "defu";
import { getSlotsChildren } from "../../utils/index.mjs";
import { useAppConfig } from "#imports";
import appConfig from "#build/app.config";
export default defineComponent({
  props: {
    name: {
      type: String,
      default: null
    },
    size: {
      type: String,
      default: null,
      validator(value) {
        return Object.keys(appConfig.ui.formGroup.size).includes(value);
      }
    },
    label: {
      type: String,
      default: null
    },
    description: {
      type: String,
      default: null
    },
    required: {
      type: Boolean,
      default: false
    },
    help: {
      type: String,
      default: null
    },
    error: {
      type: [String, Boolean],
      default: null
    },
    hint: {
      type: String,
      default: null
    },
    ui: {
      type: Object,
      default: () => appConfig.ui.formGroup
    }
  },
  setup(props, { slots }) {
    const appConfig2 = useAppConfig();
    const ui = computed(() => defu({}, props.ui, appConfig2.ui.formGroup));
    provide("form-path", props.name);
    const formErrors = inject("form-errors", null);
    const errorMessage = computed(() => {
      return props.error && typeof props.error === "string" ? props.error : formErrors?.value?.find((error) => error.path === props.name)?.message;
    });
    const children = computed(() => getSlotsChildren(slots));
    const clones = computed(() => children.value.map((node) => {
      const vProps = {};
      if (errorMessage.value) {
        vProps.oldColor = node.props?.color;
        vProps.color = "red";
      } else if (vProps.oldColor) {
        vProps.color = vProps.oldColor;
      }
      if (props.name) {
        vProps.name = props.name;
      }
      if (props.size) {
        vProps.size = props.size;
      }
      return cloneVNode(node, vProps);
    }));
    return () => h("div", { class: [ui.value.wrapper] }, [
      props.label && h("div", { class: [ui.value.label.wrapper, ui.value.size[props.size]] }, [
        h("label", { for: props.name, class: [ui.value.label.base, props.required && ui.value.label.required] }, props.label),
        props.hint && h("span", { class: [ui.value.hint] }, props.hint)
      ]),
      props.description && h("p", { class: [
        ui.value.description,
        ui.value.size[props.size]
      ] }, props.description),
      h("div", { class: [!!props.label && ui.value.container] }, [
        ...clones.value,
        errorMessage.value ? h("p", { class: [ui.value.error, ui.value.size[props.size]] }, errorMessage.value) : props.help ? h("p", { class: [ui.value.help, ui.value.size[props.size]] }, props.help) : null
      ])
    ]);
  }
});
